r-lib-error-1.0.md

```diff
@@ -0,0 +1,64 @@
+
+# Specification of r-lib errors, version 1.0.0
+
+Note: this is an experimental draft.
+
+Note: version 1.0.0 corresponds to the `rlang_error` class, as
+implemented in rlang 0.4.0.
+
+## Classes
+
+An r-lib error object always has at least classes `rlib_error`, `error`
+and `condition`. It may have other classes.
+
+## Members
+
+An `rlib_error` object has the following members:
+
+* `message`: `character(1)`, optional, the error message. If this is
+  missing then the `conditionMessage()` S3 method must be able to generate
+  the error message for the class.
+* `call`: optional, the captured call, of mode `call`.
+  Even if this is missing, the `conditionCall()` S3 method might still
+  generate the call.
+* `trace`: optional, an `rlib_trace` object, see below. 
+* `parent`: optional, pointer to a parent error, of class `error`. It
+  may also be `NULL`, which is the same as missing.
+* `version`: optional, `rlib_error` version, defaults to 1.0.0, i.e. this
+  specification.
+
+## Traces
+
+An `rlib_trace` object has the following members:
+
+* `calls`: A list of calls, each of mode `call`. This is the stack when
+  the error was thrown. `calls[[1]]` is the top of the stack.
+* `parents`: Integer vector of parent frame pointers, as in `sys.parents()`.
+* `envs`: List or character vector that contains the ids of the execution
+   environments of the stack.
+* `indices`: Frame numbers, typically just 1:n.
+* `version`: options, `r_lib` error version, defaults to 1.0.0, i.e. this
+  specification.
+
+`calls`, `parents`, `envs` and `indices` must have the same length.
+
+## Parent errors
+
+Parent errors may be arbitrary error objects. A parent error may also
+be another `rlib_error`, that includes another trace, etc.
+
+If the error has a parent error (i.e. `parent` is not `NULL`), then
+the trace of the error may only include part of the stack: from the top to
+the place where the error was thrown. The same holds for the parent error.
+(This is the rlang 0.4.0 behavior.)
+
+## Versioning
```

Comment:
What's the point of this? Can we really expect to add backward incompatible changes to an S3 class just by adding a version field?

Reply:
If somebody writes an S3 method for `rlang_error`, then that method will always work. I think we can just keep these fields the same, that is reasonable. I'll also add that our future new fields will be prefixed with `rlib`, so people are able to add new fields. E.g. `callr_error` will add `stderr`, `pid`, etc.

Reply:
Oh, wait, I have already changed this part... ignore the previous comment.

Reply:
So, originally I promised that we would always keep these fields the same, to stay backward compatible with 1.0. But @lionel- did not like that, so I removed it. 

I still think that we can make that promise, most fields are pretty basic, although we are not completely sure about `envs` (now `ids` in rlang).

Reply:
Agreed that apart from `ids`, the rest seems solid.

Reply:
I think this is really a special case of our general beliefs about versioning S3 classes, but I don't think we've written them down anywhere. Maybe it would be better to explicit flag some of the fields as experimental and encourage people not to rely on them yet?

Reply:
We rely on `ids` currently, so people need to put `ids` in the error objects they create. Anyway, we'll work out what `ids` should be next week, and then write it down here.